subsys/net/lib/lwm2m/ipso_light_control.c

```diff
@@ -0,0 +1,202 @@
+		return 0;
+	}
+
+	s32_t counter = *(s32_t *) data;
+
+	for (i = 0; i < MAX_INSTANCE_COUNT; i++) {
+		if (inst[i].obj && inst[i].obj_inst_id == obj_inst_id) {
+			if (counter == 0) {
+				on_time_offset[i] =
+					(s32_t)(k_uptime_get() / 1000);
+			}
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static struct lwm2m_engine_obj_inst *light_control_create(u16_t obj_inst_id)
+{
+	int index, i = 0;
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+			SYS_LOG_ERR("Can not create instance - "
+				    "already existing: %u", obj_inst_id);
+			return NULL;
+		}
+	}
+
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (!inst[index].obj) {
+			break;
+		}
+	}
+
+	if (index >= MAX_INSTANCE_COUNT) {
+		SYS_LOG_ERR("Can not create instance - "
+			    "no more room: %u", obj_inst_id);
+		return NULL;
+	}
+
+	/* Set default values */
+	on_off_value[index] = false;
+	dimmer_value[index] = 0;
+	on_time_value[index] = 0;
+	on_time_offset[index] = 0;
+	cumulative_active_value[index].val1 = 0;
+	cumulative_active_value[index].val2 = 0;
+	power_factor_value[index].val1 = 0;
+	power_factor_value[index].val2 = 0;
+	colour[index][0] = '\0';
+	units[index][0] = '\0';
+
+	/* initialize instance resource data */
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_ON_OFF_ID,
+		&on_off_value[index], sizeof(*on_off_value));
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_DIMMER_ID,
+		&dimmer_value[index], sizeof(*dimmer_value));
+	INIT_OBJ_RES(res[index], i, LIGHT_ON_TIME_ID, 0, NULL,
+		sizeof(*on_time_value), on_time_read_cb,
+		on_time_pre_write_cb, on_time_post_write_cb, NULL);
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_CUMULATIVE_ACTIVE_POWER_ID,
+		&cumulative_active_value[index],
+		sizeof(*cumulative_active_value));
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_POWER_FACTOR_ID,
+		&power_factor_value[index], sizeof(*power_factor_value));
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_COLOUR_ID,
+		colour[index], LIGHT_STRING_SHORT);
+	INIT_OBJ_RES_DATA(res[index], i, LIGHT_SENSOR_UNITS_ID,
+		units[index], LIGHT_STRING_SHORT);
+
+	inst[index].resources = res[index];
+	inst[index].resource_count = i;
+	SYS_LOG_DBG("Create IPSO Light Control instance: %d",
+		    obj_inst_id);
+	return &inst[index];
+}
+
+static int ipso_light_control_init(struct device *dev)
+{
+	int ret = 0;
+
+	/* Set default values */
+	memset(inst, 0, sizeof(*inst) * MAX_INSTANCE_COUNT);
+	memset(res, 0, sizeof(struct lwm2m_engine_res_inst) *
+		       MAX_INSTANCE_COUNT * LIGHT_MAX_ID);
+
+	light_control.obj_id = IPSO_OBJECT_LIGHT_CONTROL_ID;
+	light_control.fields = fields;
+	light_control.field_count = sizeof(fields) / sizeof(*fields);
+	light_control.max_instance_count = MAX_INSTANCE_COUNT;
+	light_control.create_cb = light_control_create;
+	lwm2m_register_obj(&light_control);
+
+	return ret;
+}
+
+SYS_INIT(ipso_light_control_init, APPLICATION,
```

Comment:
My only concern was that if there is some dependency to object ordering and networking, then the current proposal looks error prone. But if the overall design does not have these kind of dependency, then no worries here.